spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,502 @@
+        new SimpleRecord(3, "three"));
+
+    List<SimpleRecord> orcFileRecords = Lists.newArrayList(
+        new SimpleRecord(4, "four"),
+        new SimpleRecord(5, "five"));
+
+    List<SimpleRecord> avroFileRecords = Lists.newArrayList(
+        new SimpleRecord(6, "six"),
+        new SimpleRecord(7, "seven"));
+
+    appendData(parquetFileRecords, location, "parquet");
+    appendData(orcFileRecords, location, "orc");
+    appendData(avroFileRecords, location, "avro");
+
+    table.refresh();
+
+
+    try {
+      Dataset<Row> ds = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      List<SimpleRecord> actual = processStreamTillEnd(ds);
+      List<SimpleRecord> expected = Stream.concat(Stream.concat(
+          parquetFileRecords.stream(),
+          orcFileRecords.stream()),
+          avroFileRecords.stream())
+            .collect(Collectors.toList());
+      Assert.assertEquals(expected, actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamFromEmptyTable() throws IOException, TimeoutException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+
+    table.refresh();
+
+    try {
+      Dataset<Row> df = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+
+      List<SimpleRecord> actual = processStreamTillEnd(df);
+      Assert.assertEquals(Collections.emptyList(), actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamWithSnapshotTypeOverwriteErrorsOut() throws IOException, TimeoutException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+
+    // upgrade table to verison 2 - to facilitate creation of Snapshot of type OVERWRITE.
+    TableOperations ops = ((BaseTable) table).operations();
+    TableMetadata meta = ops.current();
+    ops.commit(meta, meta.upgradeToFormatVersion(2));
+
+    // fill table with some data
+    List<List<SimpleRecord>> dataAcrossSnapshots = getTestDataForMultipleSnapshots();
+    appendData(dataAcrossSnapshots, location);
+
+    table.refresh();
+
+    Schema deleteRowSchema = table.schema().select("data");
+    Record dataDelete = GenericRecord.create(deleteRowSchema);
+    List<Record> dataDeletes = Lists.newArrayList(
```

Comment:
I think you could just run a `DELETE FROM` or `MERGE INTO` SQL query.

Reply:
@rdblue - i wanted to do that. 
do you know how can I run Sql - when I am operating on FileLocation based tables?

Reply:
You can use a Hadoop catalog that wraps location-based tables in a catalog. Then you can point to the table directly if you want to exercise the code path that identifies tables by location.

Reply:
@rdblue - I could get `DELETE FROM` to work. When I try to use `Merge into` - I get `UnsupportedOperationException`.

Reply:
hmm. even for the `DELETE FROM` statement - I am hitting errors in the CI. This runs fine on local though.
```
org.apache.iceberg.spark.source.TestStructuredStreamingRead3 > testReadStreamWithSnapshotTypeDeleteErrorsOut[catalogName = testhive, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hive, default-namespace=default}] FAILED
    org.apache.spark.sql.AnalysisException at TestStructuredStreamingRead3.java:367
```

Reply:
@rdblue - i had to revert this change - to keep this PR green in Ci. 
since, this is a non-blocking comment & just a nicety - i am planning to this up as an independent PR - as it needs some debugging with CI. 
added it to the **What Next** section of this PR.
pl. let me know if you feel strongly for some reason - that this should block the current work to go in.

Reply:
The row-level operations require SQL extensions to run. This is fine for testing in the Spark module instead of the extensions module.